File: ingest/chunker.py
import os
import re
from pathlib import Path

import tiktoken
from dotenv import load_dotenv

load_dotenv()

CHUNK_SIZE = int(os.getenv("CHUNK_SIZE", "500"))
CHUNK_OVERLAP = int(os.getenv("CHUNK_OVERLAP", "50"))

_encoder = tiktoken.get_encoding("cl100k_base")
_HEADER_RE = re.compile(r"^#{1,3} .+", re.MULTILINE)


def _count(text: str) -> int:
    return len(_encoder.encode(text))


def chunk_markdown(filepath: str | Path) -> list[dict]:
    """
    Divide um arquivo .md em chunks semânticos.

    Retorna lista de dicts com:
      text, source_file, collection, chunk_index, type
    """
    filepath = Path(filepath)
    source_file = filepath.name
    collection = filepath.parent.name

    text = filepath.read_text(encoding="utf-8")
    paragraphs = [p.strip() for p in re.split(r"\n{2,}", text) if p.strip()]

    result: list[dict] = []
    chunk_index = 0
    current_header = ""   # último header visto
    chunk_header = ""     # header ativo quando o chunk atual começou
    buf: list[str] = []
    buf_tokens = 0

    def emit():
        nonlocal chunk_index
        if not buf:
            return
        body = "\n\n".join(buf)
        # Garante contexto do header no início do chunk
        if chunk_header and not body.startswith(chunk_header):
            full = f"{chunk_header}\n\n{body}"
        else:
            full = body
        result.append(
            {
                "text": full.strip(),
                "source_file": source_file,
                "collection": collection,
                "chunk_index": chunk_index,
                "type": "markdown",
            }
        )
        chunk_index += 1

    def rollover():
        """Mantém os últimos CHUNK_OVERLAP tokens no buffer para o próximo chunk."""
        nonlocal buf, buf_tokens
        new_buf: list[str] = []
        new_tokens = 0
        for part in reversed(buf):
            t = _count(part)
            if new_tokens + t > CHUNK_OVERLAP:
                break
            new_buf.insert(0, part)
            new_tokens += t
        buf = new_buf
        buf_tokens = new_tokens

    for para in paragraphs:
        # Rastreia o header da seção atual
        if _HEADER_RE.match(para):
            current_header = para.splitlines()[0]

        para_tokens = _count(para)

        # Parágrafo maior que CHUNK_SIZE (ex: bloco de código longo): split por linha
        if para_tokens > CHUNK_SIZE:
            if buf:
                emit()
                rollover()
                chunk_header = current_header
            for line in para.splitlines():
                line = line.strip()
                if not line:
                    continue
                line_tokens = _count(line)
                if buf_tokens + line_tokens > CHUNK_SIZE and buf:
                    emit()
                    rollover()
                    chunk_header = current_header
                buf.append(line)
                buf_tokens += line_tokens
        else:
            if buf_tokens + para_tokens > CHUNK_SIZE:
                emit()
                rollover()
                chunk_header = current_header
            buf.append(para)
            buf_tokens += para_tokens

    if buf:
        emit()

    return result

File: ingest/chunker.ts
/**
 * Chunker de documentos Markdown.
 *
 * Divide um arquivo .md em chunks semânticos respeitando o tamanho máximo
 * em palavras (aproximação de tokens) e preservando contexto de seção.
 */

import fs from "fs";
import path from "path";

export interface Chunk {
  text: string;
  sourceFile: string;
  collection: string;
  chunkIndex: number;
  type: "markdown";
}

// 1 palavra ≈ 1.3 tokens — estimativa suficiente para chunking
const WORDS_PER_TOKEN = 1.3;

function countTokens(text: string): number {
  return Math.ceil(text.split(/\s+/).filter(Boolean).length * WORDS_PER_TOKEN);
}

/**
 * Extrai o último header (# / ## / ###) visto antes de um trecho,
 * para ser prefixado no próximo chunk e manter contexto.
 */
function extractLastHeader(text: string): string {
  const lines = text.split("\n");
  let lastHeader = "";
  for (const line of lines) {
    if (/^#{1,3}\s/.test(line)) lastHeader = line;
  }
  return lastHeader;
}

export function chunkMarkdown(
  filepath: string,
  chunkSize = 500,
  chunkOverlap = 50
): Chunk[] {
  const content = fs.readFileSync(filepath, "utf-8");
  const sourceFile = path.basename(filepath);
  const collection = path.basename(path.dirname(filepath));

  // Divide em parágrafos (blocos separados por linha em branco)
  const paragraphs = content.split(/\n\n+/);

  const chunks: Chunk[] = [];
  let currentParts: string[] = [];
  let currentTokens = 0;
  let lastHeader = "";
  let chunkIndex = 0;

  const flush = () => {
    if (currentParts.length === 0) return;
    const text = currentParts.join("\n\n").trim();
    if (text) {
      chunks.push({ text, sourceFile, collection, chunkIndex, type: "markdown" });
      chunkIndex++;
    }
  };

  for (const para of paragraphs) {
    const trimmed = para.trim();
    if (!trimmed) continue;

    // Rastreia o último header para contexto
    if (/^#{1,3}\s/.test(trimmed)) lastHeader = trimmed;

    const paraTokens = countTokens(trimmed);

    // Se o parágrafo sozinho ultrapassa o limite, divide por linhas
    if (paraTokens > chunkSize) {
      flush();
      currentParts = [];
      currentTokens = 0;

      const lines = trimmed.split("\n");
      for (const line of lines) {
        const lineTokens = countTokens(line);
        if (currentTokens + lineTokens > chunkSize && currentParts.length > 0) {
          flush();
          // Começa próximo chunk com o header de contexto se existir
          currentParts = lastHeader ? [lastHeader] : [];
          currentTokens = lastHeader ? countTokens(lastHeader) : 0;
          chunkIndex++;
        }
        currentParts.push(line);
        currentTokens += lineTokens;
      }
      continue;
    }

    // Flush se adicionar este parágrafo ultrapassaria o limite
    if (currentTokens + paraTokens > chunkSize && currentParts.length > 0) {
      flush();
      // Overlap: mantém o header de contexto no próximo chunk
      currentParts = lastHeader ? [lastHeader] : [];
      currentTokens = lastHeader ? countTokens(lastHeader) : 0;
    }

    currentParts.push(trimmed);
    currentTokens += paraTokens;
  }

  flush();
  return chunks;
}
